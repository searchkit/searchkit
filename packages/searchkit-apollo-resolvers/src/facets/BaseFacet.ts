import { MixedFilter } from '../core/QueryManager'

export interface FacetOptions {
  size: number
  query: string
}

export interface BaseFacet {
  excludeOwnFilters: boolean
  getIdentifier(): string
  getLabel(): string
  getAggregation(overrides: FacetOptions): any
  getFilters(filters: Array<MixedFilter>): any
  transformResponse(response: any): any
}
