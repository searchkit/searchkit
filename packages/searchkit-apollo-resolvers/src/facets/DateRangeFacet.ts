import { DateRangeFilter } from '../core/QueryManager'
import { BaseFacet } from './BaseFacet'

interface DateRangeFacetConfig {
  id: string
  field: string
  label: string
  size?: number
  display?: 'DateRange' | string
  range: {
    min: number
    max: number
    interval: number
    min_doc_count?: number
  }
}

class DateRangeFacet implements BaseFacet {
  public excludeOwnFilters = true

  constructor(public config: DateRangeFacetConfig) {}
  getLabel(): string {
    return this.config.label
  }

  getId() {
    return this.config.id
  }

  getFilters(filters: Array<DateRangeFilter>) {
    return { range: { [this.config.field]: { gte: filters[0].dateMin, lte: filters[0].dateMax } } }
  }

  getAggregation() {
    return {}
  }

  transformResponse() {
    return {
      id: this.getId(),
      label: this.getLabel(),
      type: 'DateRangeFacet',
      display: this.config.display || 'DateRangeFacet',
      entries: []
    }
  }
}

export default DateRangeFacet
