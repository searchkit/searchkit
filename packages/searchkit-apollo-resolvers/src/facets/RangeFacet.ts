import { RangeFilter } from '../core/QueryManager'
import { BaseFacet } from './BaseFacet'

interface RangeFacetConfig {
  id: string
  field: string
  label: string
  size?: number
  display?: 'Slider' | string
  range: {
    min: number
    max: number
    interval: number
    min_doc_count?: number
  }
}

class RangeFacet implements BaseFacet {
  public excludeOwnFilters = true

  constructor(public config: RangeFacetConfig) {}
  getLabel(): string {
    return this.config.label
  }

  getId(): string {
    return this.config.id
  }

  getFilters(filters: Array<RangeFilter>) {
    return { range: { [this.config.field]: { gte: filters[0].min, lte: filters[0].max } } }
  }

  getAggregation() {
    return {
      [this.config.id]: {
        histogram: {
          field: this.config.field,
          interval: this.config.range.interval,
          min_doc_count: this.config.range.min_doc_count || 0,
          extended_bounds: { min: this.config.range.min, max: this.config.range.max }
        }
      }
    }
  }

  transformResponse(response: any) {
    return {
      id: this.getId(),
      label: this.getLabel(),
      display: this.config.display || 'RangeSliderFacet',
      type: 'RangeFacet',
      entries: response.buckets.map((entry) => ({
        id: `${this.getId()}_${entry.key}`,
        label: entry.key,
        count: entry.doc_count
      }))
    }
  }
}

export default RangeFacet
