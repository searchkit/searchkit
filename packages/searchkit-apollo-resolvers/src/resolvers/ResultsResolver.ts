import QueryManager, { MixedFilter } from '../core/QueryManager'
import SearchkitRequest from '../core/SearchkitRequest'
import BaseQuery from '../query/BaseQuery'
import { BaseFacet } from '../facets/BaseFacet'

export interface SortingOption {
  id: string
  label: string
  field: any
  defaultOption: boolean
}

export interface SearchkitConfig {
  host: string
  index: string
  sortOptions?: SortingOption[]
  hits: {
    fields: string[]
  }
  query?: BaseQuery
  facets?: Array<BaseFacet>
}

export interface ResultsResolverParameters {
  filters: Array<MixedFilter>
  query: string
}

export default (config: SearchkitConfig) => async (parent, parameters, ctx) => {
  try {
    const skConfig = {
      sortOptions: [],
      ...config
    }
    const queryManager = new QueryManager(parameters.filters, parameters.query)
    const skRequest = new SearchkitRequest(queryManager, skConfig)

    ctx.searchkit = {
      skRequest: skRequest,
      queryManager: queryManager,
      config: skConfig
    }

    return {}
  } catch (e) {
    console.log(e)
  }
}
