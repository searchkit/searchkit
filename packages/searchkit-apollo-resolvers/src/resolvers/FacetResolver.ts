import SearchkitRequest from '../core/SearchkitRequest'
import { getAggregationsFromFacets, getFacetsFromResponse } from '../core/FacetsFns'
import QueryManager from '../core/QueryManager'
import { SearchkitConfig } from './ResultsResolver'

export interface FacetResolverParameters {
  id: string
  query: string
  size: number
}

export default async (parent, parameters: FacetResolverParameters, ctx) => {
  const queryManager: QueryManager = ctx.searchkit.queryManager
  const config: SearchkitConfig = ctx.searchkit.config
  const skRequest: SearchkitRequest = ctx.searchkit.skRequest

  const facet = config.facets && config.facets.find((facet) => facet.getId() === parameters.id)
  if (!facet) return null

  try {
    const aggs = getAggregationsFromFacets(
      queryManager,
      {
        [facet.getId()]: {
          query: parameters.query,
          size: parameters.size
        }
      },
      [facet]
    )
    const response = await skRequest.search(aggs)
    const facets = getFacetsFromResponse([facet], response)

    return facets[0]
  } catch (e) {
    console.log(e)
  }
}
