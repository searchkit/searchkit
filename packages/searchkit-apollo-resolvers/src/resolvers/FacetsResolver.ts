import SearchkitRequest from '../core/SearchkitRequest'
import { getAggregationsFromFacets, getFacetsFromResponse } from '../core/FacetsFns'
import QueryManager from '../core/QueryManager'
import { SearchkitConfig } from './ResultsResolver'

export default async (parent, {}, ctx) => {
  const queryManager: QueryManager = ctx.searchkit.queryManager
  const config: SearchkitConfig = ctx.searchkit.config
  const skRequest: SearchkitRequest = ctx.searchkit.skRequest
  if (!config.facets) return null

  try {
    const aggs = getAggregationsFromFacets(queryManager, {}, config.facets)
    const results = await skRequest.search(aggs)
    const facets = getFacetsFromResponse(config.facets, results)

    return facets
  } catch (e) {
    console.log(e)
  }
}
