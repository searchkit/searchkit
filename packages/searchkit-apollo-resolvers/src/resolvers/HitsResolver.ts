import SearchkitRequest from '../core/SearchkitRequest'
import { SearchkitConfig, SortingOption } from './ResultsResolver'

export interface HitsParameters {
  page: {
    from: number
    size: number
  }
  sortBy?: string
}

function getSortOption(id, sortOptions: SortingOption[]) {
  const selectedSortOption = sortOptions.find((sortOption) => sortOption.id === id)
  if (!selectedSortOption) {
    throw new Error(`Sort Option: sorting option ${id} not found`)
  }
  return selectedSortOption
}

export default async (parent, parameters: HitsParameters, ctx) => {
  const config: SearchkitConfig = ctx.searchkit.config
  const skRequest: SearchkitRequest = ctx.searchkit.skRequest
  try {
    const from = parameters.page?.from || 0
    const size = parameters.page?.size || 10
    const defaultSortOption = config.sortOptions.find((so) => so.defaultOption)

    const chosenSortOption = parameters.sortBy
      ? getSortOption(parameters.sortBy, config.sortOptions)
      : defaultSortOption

    const { hits } = await skRequest.search({
      from: from,
      size: size,
      sort: chosenSortOption ? chosenSortOption.field : [{ _score: 'desc' }]
    })

    return {
      items: hits.hits.map((hit) => ({
        id: hit._id,
        fields: hit._source
      })),
      page: {
        total: hits.total.value,
        totalPages: Math.ceil(hits.total.value / size),
        pageNumber: from / size,
        from: from,
        size: size
      },
      sortedBy: chosenSortOption?.id
    }
  } catch (e) {
    console.log(e)
  }
}
