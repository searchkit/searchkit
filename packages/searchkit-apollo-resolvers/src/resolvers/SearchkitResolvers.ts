import ResultsResolver, { SearchkitConfig } from './ResultsResolver'
import HitsResolver from './HitsResolver'
import FacetsResolver from './FacetsResolver'
import SummaryResolver from './SummaryResolver'
import FacetResolver from './FacetResolver'

export default (config: SearchkitConfig) => ({
  ResultSet: {
    hits: HitsResolver,
    facets: FacetsResolver,
    facet: FacetResolver,
    summary: SummaryResolver
  },
  Query: {
    results: ResultsResolver(config)
  },
  FacetSet: {
    __resolveType: (e) => e.type
  },
  SelectedFilter: {
    __resolveType: (e) => e.type
  }
})
