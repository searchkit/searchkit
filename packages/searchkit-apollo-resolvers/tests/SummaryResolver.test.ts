import { SearchkitConfig } from '../src/resolvers/ResultsResolver'
import { MultiMatchQuery } from '../src'
import { RefinementSelectFacet } from '../src/facets'
import { setupTestServer, callQuery } from './support/helper'

describe('Summary Resolver', () => {
  describe('should return as expected', () => {
    const runQuery = async (gql) => {
      const response = await callQuery({ gql })
      return response
    }

    const config: SearchkitConfig = {
      host: 'http://localhost:9200',
      index: 'movies',
      hits: {
        fields: ['actors', 'writers']
      },
      query: new MultiMatchQuery({ fields: ['actors', 'writers', 'title^4', 'plot'] }),
      facets: [
        new RefinementSelectFacet({
          id: 'writers',
          field: 'writers.raw',
          label: 'Writers',
          multipleSelect: true
        }),
        new RefinementSelectFacet({ id: 'actors', field: 'actors.raw', label: 'Actors' }),
        new RefinementSelectFacet({ id: 'type', field: 'type.raw', label: 'Type' }),
        new RefinementSelectFacet({ id: 'genres', field: 'genres.raw', label: 'Genres' })
      ]
    }

    it('should return correct summary', async () => {
      setupTestServer(config)

      const gql = `
        {
          results(query: "", filters: [{ id: "writers", value: "Jeff Lindsay" }]) {
            summary {
              total
              appliedFilters {
                id
                label
                value
              }
              query
            }
            hits(page: {size: 10, from: 0 }) {
              items {
                id
              }
              page {
                total
                totalPages
                pageNumber
                from
                size
              }
            }
          }
        }
      `

      const response = await runQuery(gql)
      expect(response.body.data).toMatchSnapshot()
      expect(response.status).toEqual(200)
    })
  })
})
