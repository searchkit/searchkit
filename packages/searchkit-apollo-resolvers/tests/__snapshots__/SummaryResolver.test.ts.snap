// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Summary Resolver should return as expected should return correct summary 2`] = `
Object {
  "results": Object {
    "hits": Object {
      "items": Array [
        Object {
          "id": "tt0468569",
        },
        Object {
          "id": "tt0110912",
        },
        Object {
          "id": "tt0167260",
        },
        Object {
          "id": "tt0114369",
        },
        Object {
          "id": "tt0073486",
        },
        Object {
          "id": "tt0172495",
        },
        Object {
          "id": "tt1520211",
        },
        Object {
          "id": "tt0338013",
        },
        Object {
          "id": "tt0078788",
        },
        Object {
          "id": "tt0180093",
        },
      ],
      "page": Object {
        "from": 0,
        "pageNumber": 0,
        "size": 10,
        "total": 4162,
        "totalPages": 417,
      },
    },
    "summary": Object {
      "appliedFilters": Array [
        Object {
          "display": "ListFacet",
          "id": "actors_Jeff Lindsay",
          "identifier": "actors",
          "label": "Actors",
          "value": "Jeff Lindsay",
        },
        Object {
          "display": "RangeSliderFacet",
          "id": "imdbrating_0_50",
          "identifier": "imdbrating",
          "label": "IMDB Rating",
          "max": 50,
          "min": 0,
        },
        Object {
          "dateMax": "2020-12-18T00:00:00.000Z",
          "dateMin": "2012-12-18T00:00:00.000Z",
          "display": "DateRangeFacet",
          "id": "released_2012-12-18T00:00:00.000Z_2020-12-18T00:00:00.000Z",
          "identifier": "released",
          "label": "Release",
        },
      ],
      "query": "",
      "sortOptions": Array [
        Object {
          "id": "relevance",
          "label": "Relevance",
        },
        Object {
          "id": "released",
          "label": "Recent Releases",
        },
      ],
      "total": 4162,
    },
  },
}
`;
